### PR TITLE
Use block scope thread_local static for NSProgress progress stack

### PR DESCRIPTION
**Pull request by ehren**
per comments in #1872. Works around issue with file_scope thread_local in current clang. Leak check unit test added.

### CODE REVIEW

File: tests/unittests/Foundation/NSProgressTests.mm [hunk 1]
```diff
@@ -583,3 +583,16 @@ - (void)addParentTo:(NSProgress*)child {
...
+    NSProgress* progress = [NSProgress progressWithTotalUnitCount:100];
+
+    std::thread t([&progress]() {
+        [progress becomeCurrentWithPendingUnitCount:50];
+    });
+
+    t.join();
+
+    // Progress unfinished upon thread exit. Ensure we're not keeping progress alive:
+    EXPECT_EQ(1, [progress retainCount]);
```
Comment (ehren) [inline tests/unittests/Foundation/NSProgressTests.mm]: Not the biggest fan of asserting on the retain count everywhere in tests.... but a conversion to ARC would be straightforward.

File: Frameworks/Foundation/NSProgress.mm [hunk 2]
```diff
@@ -46,16 +46,15 @@
...
-
 // Returns the stack of NSProgress objects that have becomeCurrent on the current thread, initializing it if necessary
-static decltype(s_currentProgressStack) & _getProgressStackForCurrentThread() {
-    if (!s_currentProgressStack) {
-        s_currentProgressStack = std::make_shared<std::stack<CurrentProgress>>();
+static std::shared_ptr<std::stack<CurrentProgress>> & _getProgressStackForCurrentThread() {
+    thread_local static std::shared_ptr<std::stack<CurrentProgress>> tlsCurrentProgressStack;
+
+    if (!tlsCurrentProgressStack) {
+        tlsCurrentProgressStack = std::make_shared<std::stack<CurrentProgress>>();
```
Comment (rajsesh-msft) [inline Frameworks/Foundation/NSProgress.mm]: These three lines could be collapsed into thread_local static std::shared_ptr<std::stack<CurrentProgress>> tlsCurrentProgressStack = std::make_shared<std::stack<CurrentProgress>>()

Comment (ehren) [inline Frameworks/Foundation/NSProgress.mm]: @rajsesh-msft fixed. Also, it's not expressly allowed by the style guide but what about using auto& / return type deduction to replace the removed use of decltype?

Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSProgress.mm]: @ehren we do that in a lot of places where the type is expressed fully on the righthand side :)

Comment (ehren) [inline Frameworks/Foundation/NSProgress.mm]: @DHowett-MSFT cool, switched in latest.

Comment (ehren) [inline Frameworks/Foundation/NSProgress.mm]: the double auto might be skirting the rules a bit... no prob either way to switch.

**Comment by rajsesh-msft** [comment]
thank you @ehren!

**Pull request merged**